redfish/collector.go

```diff
@@ -0,0 +1,241 @@
+package redfish
+
+import (
+	"context"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/cybozu-go/log"
+	"github.com/cybozu-go/setup-hw/config"
+	"github.com/cybozu-go/setup-hw/gabs"
+	"github.com/prometheus/client_golang/prometheus"
+	yaml "gopkg.in/yaml.v2"
+)
+
+const namespace = "hw"
+
+// Collector implements prometheus.Collector interface.
+type Collector struct {
+	rules  []convertRule
+	client *client
+	cache  *cache
+}
+
+// CollectorConfig is a set of configurations for Collector.
+type CollectorConfig struct {
+	AddressConfig *config.AddressConfig
+	Port          string
+	UserConfig    *config.UserConfig
+	Rule          []byte
+}
+
+// NewCollector returns a new instance of Collector.
+func NewCollector(cc *CollectorConfig) (*Collector, error) {
+	cache := &cache{dataMap: make(dataMap)}
+	client, err := newClient(cc, cache)
+	if err != nil {
+		return nil, err
+	}
+
+	var rules []convertRule
+	err = yaml.Unmarshal(cc.Rule, &rules)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, rule := range rules {
+		if err := rule.Validate(); err != nil {
+			return nil, err
+		}
+	}
+
+	return &Collector{rules: rules, client: client, cache: cache}, nil
+}
+
+// Describe does nothing for now.
+func (c Collector) Describe(ch chan<- *prometheus.Desc) {
```

Comment:
Please describe the rationale for not implementing this.

Reply:
Mu-mu-mu...
This is corner-cutting.
I should implement this by processing conversion rules.

Moreover, the usage of Desc objects in `Collect()` is wrong, though it is copied from dellhw_exporter.
I'll fix it too.